### ISSUE TITLE
Support fallible eq and hasher in raw API

### ISSUE THREAD

**Issue by udoprog** [rust-lang/hashbrown#456]
Hi!

I want to use the raw API of `hashbrown` as the underlying table implementation in [Rune](https://github.com/udoprog/rune-rs/rune).

One aspect which has proven to be awkward is that the hasher I need to pass into `RawTable::find_or_find_insert_slot` is infallible, which works well for a statically typed language, but less so in Rune where I want to store and deal with dynamically typed entries directly.

> Note: I've implemented a prototype of this in my [`raw-infallible-context` branch][branch].

What I need to do is something like this:

```rust
let mut table: RawTable<(Value, Value)> = todo!();
let entry: (Value, Value) = todo!();

// NB: might fail if the key can't be hashed.
let hash = key.hash()?;

// Comparison might fail.
let eq = |other: &(Value, Value)| -> Result<bool, MyError> { Value::eq(&entry.0, &other.0) };
// Hashing might fail.
let hasher = |key: &(Value, Value)| -> Result<u64, MyError> { Value::hash(&key.0) }

let result = table.find_or_find_insert_slot(hash, eq, hasher);
// deal with my custom result
```

The change needed would be fairly straightforward ([I think][branch]) but a bit tedious, what I think is needed is to refactor the API into a pair of methods like this (ignore naming):

```rust
use std::convert::Infallible;

#[inline(always)]
fn into_ok<T>(result: Result<T, Infallible>) -> T {
    match result {
        Ok(value) => value,
        Err(error) => match error {},
    }
}

#[inline(always)]
fn infallible_eq<T>(mut f: impl FnMut(&T) -> bool) -> impl FnMut(&T) -> Result<bool, Infallible> {
    move |value| Ok::<_, Infallible>(f(value))
}

#[inline(always)]
fn infallible_hasher<T>(f: impl Fn(&T) -> u64) -> impl Fn(&T) -> Result<u64, Infallible> {
    move |value| Ok::<_, Infallible>(f(value))
}

impl RawTable {
    /// Backwards compatibility
    #[inline(always)]
    pub fn find_or_find_insert_slot(
        &mut self,
        hash: u64,
        eq: impl FnMut(&T) -> bool,
        hasher: impl Fn(&T) ->  u64,
    ) -> Result<Bucket<T>, InsertSlot> {
        into_ok(self.find_or_find_insert_slot2(hash, infallible_eq(eq), infallible_hasher(hasher)))
    }

    /// New method
    pub fn find_or_find_insert_slot2<E>(
        &mut self,
        hash: u64,
        mut eq: impl FnMut(&T) -> Result<bool, E>,
        hasher: impl Fn(&T) -> Result<u64, E>,
    ) -> Result<Result<Bucket<T>, InsertSlot>, E> {
        todo!()
    }
}
```

The signatures leave something to be desired and such a change would have to be propagated through the rest of the API, [the majority if which have proven to be internal][branch]. I would also hope that the compiler can easily prove the infallible variants are no-ops but this is something that should probably be investigated.

[branch]: https://github.com/rust-lang/hashbrown/compare/master...udoprog:hashbrown:raw-infallible-context?expand=1

**Comment by Amanieu** [comment]
I don't think supporting fallible hasher is worth adding this level of complexity to the API. In most cases this isn't needed, and you can always emulate it by returning a hash value of 0 and recording the error in an `Option` that you check after the hash table operation is complete.

However I do think there is value in `find_or_find_insert_slot_with_context`: it is very common to want both `eq` and `hasher` to have access to the same object. It just isn't an issue in practice since mutable access is not usually required (shared access is sufficient).

**Comment by udoprog** [comment]
> I don't think supporting fallible hasher is worth adding this level of complexity to the API. In most cases this isn't needed, and you can always emulate it by returning a hash value of 0 and recording the error in an `Option` that you check after the hash table operation is complete.

I considered this initially, but would this not result in potentially ill-defined downstream behaviors in fallible cases (Same for `eq`)? Preferably I'd want to ensure that any errors prevent the modifications from taking place to achieve my version of exception safety.

**Comment by Amanieu** [comment]
`hasher` absolutely cannot be allowed to fail, we rely on this when rehashing the table. See `rehash_in_place` where if the hasher panics then we drop all elements in the table. If this is not something that you can guarantee then you may want to consider caching hash values separately so they are always available.

`eq` failing is less serious since you can always just return `false`, and it doesn't modify the table.

**Comment by udoprog** [comment]
> `hasher` absolutely cannot be allowed to fail, we rely on this when rehashing the table. See `rehash_in_place` where if the hasher panics then we drop all elements in the table.

Right, so the behavior I'd want to emulate is what would happen if the hash implementation panics today (which is how I believe this PR works). Think of this error propagation as my own "structured panics".

> If this is not something that you can guarantee then you may want to consider caching hash values separately so they are always available.

What do you mean? Are raw Tables not exception safe?

> `eq` failing is less serious since you can always just return `false`, and it doesn't modify the table.

Doesn't that depend on where it fails? If it fails during insertion and returns `false` wouldn't it insert another bucket since two values are not considered eq but have the same hash?

**Comment by udoprog** [comment]
I couldn't produce a modification for a failing `eq`, but here is one which is avoided for a failing `hash`:

```rust
use hashbrown::raw::RawTable;

fn main() {
    let mut table = RawTable::<()>::new();

    unsafe {
        // Insert three elements (with the same hash) to fill up initial capacity:
        for _ in 0..3u64 {
            let Err(slot) = table.find_or_find_insert_slot(0, |_| false, |_| 0) else {
                panic!();
            };

            table.insert_in_slot(0, slot, ());
        }

        dbg!(table.capacity());

        // Table does not grow when `hasher` fails like this:
        let Err(()) = table.find_or_find_insert_slot_with(&mut (), 0, |_, _| Ok(false), |_, _| Err(())) else {
            panic!();
        };

        dbg!(table.capacity());

        // Table does grow when a dummy value is returned instead:
        let Err(..) = table.find_or_find_insert_slot(0, |_| false, |_| 0) else {
            panic!();
        };

        dbg!(table.capacity());
    }
}
```

**Comment by Amanieu** [comment]
> What do you mean? Are raw tables not exception safe?

Not completely: because we don't store the full hash in the table, we rely on the hasher when rehashing the table in-place. If the hasher panics then we have no way of knowing where to place elements in the table. It's exception-sense in the sense that you won't get UB, but you may lose elements in the table.

https://github.com/rust-lang/hashbrown/blob/bb6521e2cbf218b7d3070b58ced30907d248d376/src/raw/mod.rs#L2277-L2280

> Doesn't that depend on where it fails? If it fails during insertion (e.g. `find_or_find_insert_slot`) and `eq` returns `false` wouldn't the table allocate another bucket since two values are not considered eq but have the same hash?

`find_or_find_insert_slot` first calls `reserve(1)` to ensure there is enough space in the table in case an insertion is needed, before doing anything. This function doesn't actually perform any insertion into the table, that is done by `insert_in_slot`.

**Comment by udoprog** [comment]
Right. Thanks for the clarity. So that is what I gathered from reading the implementation as well. The only extent I expect the table to be exception safe is that it doesn't permit anything memory unsafe to happen if we do end up catching a panic. But the exact end state is not guaranteed to be the same as prior to the attempted modification.

I am still interested to ensure that my "errors" exhibit the same behaviour as a caught Rust panic, mainly because I want to provide as much feature parity between Rust and Rune as possible. So I still want to provide fallibility to the methods.

It's nice that you're interested in passing the context around though (possibly the bigger change). But until fallible methods are supported or the exact behavior of returning "dummy values" is understood and tested* I'll probably end up maintaining a private fork of the raw table with those minor changes.

*: To the degree that it's possible to emulate the behavior of an equivalent Rust panic.
